### ISSUE TITLE
Git worktree folders shouldn't collapse

### ISSUE THREAD

**Issue by berkus** [zed-industries/zed#19847]
### Summary

My project has some git worktrees under the `.worktrees` folder in the project. 

All the file trees under this `.worktrees` folder are sometimes collapsed by zed.

<img width="287" alt="image" src="https://github.com/user-attachments/assets/32186d2f-6fea-4a0d-8810-dc0ba7dbeb6c">

I expand it:

<img width="288" alt="image" src="https://github.com/user-attachments/assets/b5b17a0f-dbfd-4390-9f22-57841d6ec381">

And after some time is collapses itself again.

If I open `.worktrees/feat/higher-half-kernel` as a separate project, it also automatically collapses:

<img width="288" alt="image" src="https://github.com/user-attachments/assets/f97caf33-b100-4388-8b8f-83aa2a2bc6ee">

That's not very convenient, because it loses all context periodically.

### Environment

Zed: v0.158.2 (Zed)
OS: macOS 15.1.0
Memory: 24 GiB
Architecture: aarch64

**Comment by notpeter** [comment]
Can you see if you can come up with some steps or an example repository which reproduces this issue?  I don't use [`git worktree`](https://git-scm.com/docs/git-worktree), but would like to be able to understand what's happening.

My blind guess is that when switching worktrees git cleans everything up (emptying the directory) before recreating all the files again.  Zed is watching those files/directories which get removed but does not attempt to rewatch the newly created files/directories at the same paths.  But that's just a guess.

**Comment by berkus** [comment]
> My blind guess is that when switching worktrees git cleans everything up (emptying the directory) before recreating all the files again.

No, worktrees are just permanent checkouts from the same bare repo into the subfolder, but I think you're right that this is somehow related to git ops - I noticed that this most often happens after I do some git manipulations in a separate window, so there are some file update events.

Couldn't reproduce this in a clean empty project right now, will look for more patterns and try to reproduce the exact scenario.

**Comment by berkus** [comment]
Yes, this still happens.

**Comment by notpeter** [comment]
This would certainly be easier to fix if anyone were able to come up with specific steps to reproduce this behavior.

My gut guess is that there is a folder being renamed into place so that our filesystem watcher is notifying of a delete event (remove folder from project panel tree) followed by a creation event (add folder to project panel tree) and the open/fold status is getting discarded as it was attached to the now deleted item and default to the closed folder.

See also:
- https://github.com/zed-industries/zed/discussions/26084

**Comment by berkus** [comment]
1. git clone https://github.com/metta-systems/vesper
2. open `vesper` in zed
3. git worktree add .worktrees/feat/parse-dtb feat/parse-dtb
4. git worktree add .worktrees/fix/remove-static-ub fix/remove-static-ub
5. open .worktrees/feat/parse-dtb
6. open any file in this worktree
7. see it is highlighted in the file navigator on the left
8. keep editing and making git commits
9. notice the file navigator changes and the whole .worktrees subtree is collapsed

**Comment by notpeter** [comment]
I have attempted to reproduce but was unsuccessful.
```sh
cd /tmp
git clone https://github.com/metta-systems/vesper
cd vesper
git worktree add .worktrees/feat/parse-dtb feat/parse-dtb
git worktree add .worktrees/fix/remove-static-ub fix/remove-static-ub
zedn . .worktrees/feat/parse-dtb/README.md
```

I then make edits/commits and have been unable to trigger the behavior.
Please provide additional explicit steps/commands so I can reproduce the behavior.

**Comment by berkus** [comment]
Press Cmd+J and do some git operations.

touch newfile
git add newfile
git commit -m "added new file"
gitup

Do some ops in gitup perhaps. If your suspicion is correct the trigger is
some fs change, so you need to make git do some fs changes.


Sent via Superhuman ***@***.***>

**Comment by notpeter** [comment]
I have attempted to perform git operations (commit, reset, checkout) but was not able to trigger the behavior you are seeing.  I totally believe you that it's acting poorly, I just need to come up with the right steps so I can reproduce and validate any fix we attempt to implement.

**Comment by berkus** [comment]
I will wake up in some 10 hours and work on a reproducible scenario. Are
you using the latest git version of zed for reproducing this?

Sent via Superhuman ***@***.***>
